### PR TITLE
Cleanup of unnecessary includes, plus other cleanups

### PR DESCRIPTION
**Pull request by tautschnig**
This is just cleanup work to reduce the number of includes where those includes are actually not necessary. This reduces the number of dependencies and thus the number of translation units that need to be rebuilt upon changes.

In my testing, I have not observed noteworthy differences in build time for a complete re-build (which would have been another desirable side effect).

### CODE REVIEW

**Comment by kroening** [comment]
compile.cpp wants util/unicode.h for widen()

File: src/goto-programs/interpreter_evaluate.cpp [hunk 1]
```diff
@@ -389,14 +389,11 @@ void interpretert::evaluate(
...
-      const char *str=value.c_str();
-      std::size_t length=strlen(str)+1;
+      const std::string value = id2string(to_constant_expr(expr).get_value());
       if(show)
         warning() << "string decoding not fully implemented "
-                  << length << eom;
-      mp_integer tmp = get_string_container()[id2string(value)];
-      dest.push_back(tmp);
+                  << value.size() + 1 << eom;
+      dest.push_back(get_string_container()[value]);
```
Comment (smowton) [inline src/goto-programs/interpreter_evaluate.cpp:390]: While you're messing with this, pretty sure `get_string_container()[value]` -> `to_constant_expr(expr).get_value().get_no()` would work?

File: src/goto-programs/interpreter_evaluate.cpp [hunk 2]
```diff
@@ -389,14 +389,11 @@ void interpretert::evaluate(
     }
     else if(expr.type().id()==ID_string)
     {
-      irep_idt value=to_constant_expr(expr).get_value();
-      const char *str=value.c_str();
-      std::size_t length=strlen(str)+1;
+      const std::string value = id2string(to_constant_expr(expr).get_value());
```
Comment (smowton) [inline src/goto-programs/interpreter_evaluate.cpp]: Could take by const ref

**Review by smowton — APPROVED** [review]
Looks good, one function could be improved while you're editing it anyway, and a few of these commits should surely have been different PRs ;)

File: src/goto-programs/interpreter_evaluate.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/interpreter_evaluate.cpp:390]: I believe that was the previous implementation, but that would not/did not work with `-DUSE_STD_STRING`. That actually was 25ffad4.

**Comment by tautschnig** [comment]
> [...] and a few of these commits should surely have been different PRs ;)

Thank you for updating the title! Indeed some of this could be factored out, but would then become a pre-dependency. If requested I'll of course factor things out.

File: src/goto-programs/string_abstraction.cpp [hunk 3]
```diff
@@ -759,7 +759,7 @@ bool string_abstractiont::build(const exprt &object, exprt &dest, bool write)
 
   if(object.id()==ID_string_constant)
   {
-    mp_integer str_len=strlen(object.get(ID_value).c_str());
+    const std::size_t str_len = id2string(object.get(ID_value)).size();
     return build_symbol_constant(str_len, str_len+1, dest);
```
Comment (kroening) [inline src/goto-programs/string_abstraction.cpp]: Oh, I believe to remember that this was intentional. Think of the case of a string literal with a '\0' in it.

Comment (tautschnig) [reply src/goto-programs/string_abstraction.cpp]: Thanks! I have now made this explicit.

**Pull request merged**